docs/design.md

```diff
@@ -0,0 +1,508 @@
+        version: 0.1.0
+        source: https://galaxy.ansible.com
+```
+
+## Supported Ansible Contents
+
+Ansible provider supports running different types of Ansible contents using `AnsibleRun`, including roles and playbooks. You can not define roles and playbooks in the same `AnsibleRun` resource. They are mutually exclusive.
+
+You have already seen how to run Ansible role and inline playbook. Here is an example to run an Ansible playbook that is included in a collection, using `spec.forProvider.playbook`:
+
+```yaml
+apiVersion: ansible.crossplane.io/v1alpha1
+kind: AnsibleRun
+metadata:
+  name: remote-example
+spec:
+  forProvider:
+    playbook: sample_namespace.sample_collection.sample_playbook
+  providerConfigRef:
+    name: provider-config-example
+```
+
+If multiple playbooks are listed using `spec.forProvider.playbooks`, they will be run sequencially one after another. At the time of writing this document, only single playbook execution is supported.
+
+```yaml
+apiVersion: ansible.crossplane.io/v1alpha1
+kind: AnsibleRun
+metadata:
+  name: remote-example
+spec:
+  forProvider:
+    playbooks:
+      - sample_namespace.sample_collection.sample_playbook
+      - sample_namespace.sample_collection.another_sample_playbook
+  providerConfigRef:
+    name: provider-config-example
+```
+
+The collection that includes the playbook is defined as a requirement in `ProviderConfig` and  retrieved from a remote place.
+
+When using inline playbook, besides builtin Ansible modules, you can also include roles or playbooks from collections that are defined as requirements in `ProviderConfig` and retrieved from remote place. For example, below inline playbook is to call `nginx` role in `nginxinc.nginx_core` collection:
+
+```yaml
+apiVersion: ansible.crossplane.io/v1alpha1
+kind: AnsibleRun
+metadata:
+  name: inline-example
+spec:
+  forProvider:
+    playbookInline: |
+      ---
+      - hosts: all
+        collections:
+          - nginxinc.nginx_core
+        tasks:
+          - name: Install NGINX
+            include_role:
+              name: nginx
+  providerConfigRef:
+    name: provider-config-example
+```
+
+The corresponding collection is defined as requirement in `ProviderConfig`:
+
+```yaml
+apiVersion: ansible.crossplane.io/v1alpha1
+kind: ProviderConfig
+metadata:
+  name: example
+spec:
+  requirements: |
+    ---
+    collections:
+      - name: nginxinc.nginx_core
+        version: 0.5.0
+```
+
+## Passing Variables
+
+Ansible uses variables to manage differences among systems on which Ansible operates, so it can run roles or playbooks on multiple systems using single command. Ansible provider allows you to pass those differences into Ansible run through `vars` field when define `AnsibleRun` resource. 
+
+Here is an example to define some variables including a simple variable, a list variable, and a dictionary variable. These variables will be passed into Ansible role `sample_namespace.sample_role` for it to consume:
+
```

Comment:
Do you mean facts vs. vars? I thought that facts do not have be exposed in `AnsibleRun` CR, of course they can be referenced inside Ansible roles or playbooks. Did I miss something?

Reply:
this can facts or other configuration vars , for example: 

- `ANSIBLE_ROLE_PATH`: https://github.com/multicloudlab/crossplane-provider-ansible/blob/main/internal/ansible/ansible.go#L34
- `ANSIBLE_COLLECTION_PATH` : https://github.com/multicloudlab/crossplane-provider-ansible/blob/main/internal/ansible/ansible.go#L36

Since we are giving the user the ability to load their roles/playbook at runtime, we need to allow them to manage their resources by exposing some native ansible environment variables, what I call configuration variables

So, I think we should support both `facts` and `vars`

Reply:
ok, so, these are really environment variables, at the concept level, neither `facts` nor `vars`, and they are predefined environment variables used by Ansible specifically. But I can see from the implementation at the moment, it is using `vars` at: https://github.com/multicloudlab/crossplane-provider-ansible/blob/main/internal/ansible/ansible.go#L263, https://github.com/multicloudlab/crossplane-provider-ansible/blob/main/internal/ansible/ansible.go#L265 to support these environment variables. So, IIUC, we use `var` to cover all, and we can just add notes in this spec to tell readers that `vars`, at least for current releases, can also be used to specify some predefined "environment variables" that will be consumed by Ansible. Does it make sense?

Reply:
@morningspace  may be we can move the concept level vars to the providerConfig resource. WDYT ?

Reply:
> may be we can move the concept level vars to the providerConfig resource.
yes, we can put that to ProviderConfig, as long as these concept level vars are the same across AnsibleRuns. Otherwise, it will be remained at AnsibleRun level.